File: two_sum.cxx
class Solution {
public:
    #if 0
    //brute force
    vector<int> twoSum(vector<int>& nums, int target) {
        for(int i=0; i<nums.size(); ++i) {
            int value = target - nums[i];
            for(int j=i; ++j<nums.size();) {
                if(nums[j]==value) {
                    return std::vector<int>{i,j};
                } 
            }  
        }
        return {};
    }
    #endif

    //hash table, one pass
    vector<int> twoSum(vector<int>& nums, int target) {
        std::unordered_map<int,int> m0;
        for(int i=0; i<nums.size(); ++i) {
            int value = target - nums[i];
            if(m0.count(value)) {
                return {m0[value],i};
            }
            m0[nums[i]]=i;
        }
        return {};
    }

File: two_sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        d0 = {}
        for i in range(len(nums)):
            val = target - nums[i]
            if val in d0:
                return [d0[val],i]    
            d0[nums[i]]=i
        return []
